torch/nn/parallel/distributed.py

```diff
@@ -1085,10 +1168,21 @@ def _run_ddp_forward(self, *inputs, **kwargs):
             with self._inside_ddp_forward():
                 return self.module(*inputs, **kwargs)
 
+    def _clear_grad_buffer(self):
+        if self._grad_buffer is not None:
+            for index, param in enumerate(self._delay_all_reduce_params):
+                param.grad = self._grad_views[index]
+            self._grad_buffer.zero_()
+
     def forward(self, *inputs, **kwargs):
         with torch.autograd.profiler.record_function(
             "DistributedDataParallel.forward"
         ):
+            if self._use_vanilla_module:
+                output = self.module.forward(*inputs, **kwargs)
+                self._clear_grad_buffer()
+                return output
```

Comment:
how about the parameters that are not managed by _delay_all_reduce_params? they should be managed by DDP reducer, for these parameters, we still need to execute the below codes for preparing DDP reducer?

Reply:
Yes. In the case where there are still params that are not managed by _delay_all_reduce_params, the flag _use_vanilla_module is False.